dl1_data_handler/writer.py

```diff
@@ -742,3 +751,63 @@ def _process_data(self, file_list, output_filename):
                     # Reset event count and increment file count
                     event_count = 0
                     output_file_count += 1
+
+
+def gain_selection(waveform, image, peakpos, cam_id, threshold):
+    """
+    Based on the waveform and threshold, select the proper gain for each pixel.
+    By default, the channel 0 is kept.
+    If a pixel is saturated (value > threshold in the waveform), the channel 1 is used.
+
+    Parameters
+    ----------
+    waveform: array of waveforms of the events
+    image: array of calibrated pixel charges
+    peakpos: array of pixel peak positions
+    cam_id: str
+    threshold: int threshold to change form high gain to low gain
+
+    Returns
+    -------
+    combined_image, combined_peakpos: `(numpy.array, numpy.array)`
+        combined_image.shape = image.shape[1]
+    """
+
+    assert image.shape[0] == 2
+
+    gainsel = ThresholdGainSelector(select_by_sample=True)
```

Comment:
This is surprisingly very expensive in terms of time, I think we should try to just initialize this once, rather than once per event, as it does now.